src/plugins/core/data_validation.ts

```diff
@@ -287,10 +288,17 @@ export class DataValidationPlugin
     const evaluator = dataValidationEvaluatorRegistry.get(criterion.type);
     if (
       criterion.values.some((value) => {
-        if (value.startsWith("=")) {
-          return evaluator.allowedValues === "onlyLiterals";
-        } else if (evaluator.allowedValues === "onlyFormulas") {
+        if (!value.startsWith("=") && evaluator.allowedValues === "onlyFormulas") {
           return true;
+        }
+        if (value.startsWith("=")) {
+          if (evaluator.allowedValues === "onlyLiterals") {
+            return true;
+          }
+          if (compile(value).isBadExpression) {
+            return true;
+          }
+          return false;
```

Comment:
This closure is quite difficult to read to me. I propose a slightly simplified version and I extracted the closure to a variable with a proper name. Still not super easy to read though
```ts
const isInvalid = (value: string) => {
  if (evaluator.allowedValues === "onlyFormulas" && !value.startsWith("=")) {
    return true;
  }
  if (value.startsWith("=")) {
    return evaluator.allowedValues === "onlyLiterals" || compile(value).isBadExpression;
  }
  return !evaluator.isCriterionValueValid(value);
}
if (criterion.values.some(isInvalid)) {
  return CommandResult.InvalidDataValidationCriterionValue;
}
```